trio/_core/_io_windows.py

```diff
@@ -97,219 +222,280 @@ class CompletionKeyEventInfo:
+                _check(kernel32.CloseHandle(iocp))
+        finally:
+            if self._afd is not None:
+                afd = self._afd
+                self._afd = None
+                _check(kernel32.CloseHandle(afd))
+
+    def __del__(self):
+        self.close()
 
     def statistics(self):
+        tasks_waiting_read = 0
+        tasks_waiting_write = 0
+        for waiter in self._afd_waiters.values():
+            if waiter.read_task is not None:
+                tasks_waiting_read += 1
+            if waiter.write_task is not None:
+                tasks_waiting_write += 1
         return _WindowsStatistics(
+            tasks_waiting_read=tasks_waiting_read,
+            tasks_waiting_write=tasks_waiting_write,
             tasks_waiting_overlapped=len(self._overlapped_waiters),
             completion_key_monitors=len(self._completion_key_queues),
-            tasks_waiting_socket_readable=len(self._socket_waiters["read"]),
-            tasks_waiting_socket_writable=len(self._socket_waiters["write"]),
-            iocp_backlog=len(self._iocp_queue),
         )
 
-    def close(self):
-        if not self._closed:
-            self._closed = True
-            _check(kernel32.CloseHandle(self._iocp))
-            if self._iocp_thread is not None:
-                self._iocp_thread.join()
-            self._main_thread_waker.close()
-
-    def __del__(self):
-        # Need to make sure we clean up self._iocp (raw handle) and the IOCP
-        # thread.
-        self.close()
-
     def handle_io(self, timeout):
-        # Step 0: the first time through, initialize the IOCP thread
-        if self._iocp_thread is None:
-            # The rare non-daemonic thread -- close() should always be called,
-            # even on error paths, and we want to join it there.
-            self._iocp_thread = threading.Thread(
-                target=self._iocp_thread_fn, name="trio-IOCP"
+        received = ffi.new("PULONG")
+        milliseconds = round(1000 * timeout)
+        if timeout > 0 and milliseconds == 0:
+            milliseconds = 1
+        try:
+            _check(
+                kernel32.GetQueuedCompletionStatusEx(
+                    self._iocp, self._events, MAX_EVENTS, received,
+                    milliseconds, 0
+                )
+            )
+        except OSError as exc:
+            if exc.winerror != ErrorCodes.WAIT_TIMEOUT:  # pragma: no cover
+                raise
+            return
+        for i in range(received[0]):
+            entry = self._events[i]
+            if entry.lpCompletionKey == CKeys.AFD_POLL:
+                lpo = entry.lpOverlapped
+                op = self._afd_ops.pop(lpo)
+                waiters = op.waiters
+                if waiters.current_op is not op:
+                    # Stale op, nothing to do
+                    pass
+                else:
+                    waiters.current_op = None
+                    # I don't think this can happen, so if it does let's crash
```

Comment:
Any reason not to raise it as the result of `wait_readable` or `wait_writable`? It's hard to track down what tasks were doing when a TrioInternalError occurred (although I know we're hoping to fix that more generally at some point).

Reply:
I thought about this a lot and made an attempt at implementing it, but it turns out to be non-trivial to wake everyone up here while keeping our internal state consistent. And that's code that we can't even test, because we don't know any way to trigger this error. Also, since we have no idea how we could hit this error, it's entirely possible that if it happens, things will be more generally broken, and trying to resume the scheduler will create a deadlock or something. So, I think I'd rather leave it as a `TrioInternalError` crash for now. (Fortunately I'm pretty sure exceptions from inside `handle_io` do come out directly from `trio.run`.)

Reply:
Works for me!